compparallel.go

```diff
@@ -0,0 +1,267 @@
+package routinghelpers
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/routing"
+)
+
+var _ routing.Routing = &Parallel{}
+
+type ComposableParallel struct {
+	routers []*ParallelRouter
+}
+
+func NewComposableParallel(routers []*ParallelRouter) *ComposableParallel {
+	return &ComposableParallel{
+		routers: routers,
+	}
+}
+
+func (r *ComposableParallel) Provide(ctx context.Context, cid cid.Cid, provide bool) error {
+	var wg sync.WaitGroup
+	errCh := make(chan error)
+	closeCh := make(chan bool)
+	for _, r := range r.routers {
+		r := r
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			select {
+			case <-ctx.Done():
+				if ctx.Err() != nil && !r.IgnoreError {
+					errCh <- ctx.Err()
+				}
+			case <-time.After(r.ExecuteAfter):
+				ctx, cancel := context.WithTimeout(ctx, r.Timeout)
+				defer cancel()
+				err := r.Router.Provide(ctx, cid, provide)
+				if err != nil &&
+					!errors.Is(err, routing.ErrNotSupported) &&
+					!r.IgnoreError {
+					errCh <- err
+				}
+			case <-closeCh:
+				return
+			}
+		}()
+	}
+
+	go func() {
+		wg.Wait()
+		close(closeCh)
+	}()
+
+	var errOut error
+	select {
+	case err := <-errCh:
+		errOut = multierror.Append(errOut, err)
+	case <-closeCh:
+	}
```

Comment:
Shouldn't this select be in a loop? I think if there are multiple errors then this will leak goroutines, as they will be stuck trying to write to errCh.

Reply:
right, fixed.